pkg/operator/loglevel/controller.go

```diff
@@ -0,0 +1,128 @@
+func Add(mgr manager.Manager, kubeConfig string) error {
+	r := newReconciler(mgr)
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CloudCredential object and reconcile the loglevel changes
+	err = c.Watch(
+		&source.Kind{Type: &operatorv1.CloudCredential{}},
+		&handler.EnqueueRequestForObject{})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	r := &ReconcileCloudCredentialConfig{
+		Client: mgr.GetClient(),
+	}
+	status.AddHandler(controllerName, r)
+
+	return r
+}
+
+var _ reconcile.Reconciler = &ReconcileCloudCredentialConfig{}
+
+// ReconcileCloudCredentialConfig reconciles a CredentialConfig object
+type ReconcileCloudCredentialConfig struct {
+	client.Client
+}
+
+// Reconcile reads the state of the cluster for a CredentialConfig object and
+// makes changes based on the state read and what is in the CredentialConfig.LogLevel
+// Automatically generate RBAC rules to allow the Controller to read and write required types.
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=cloudcredential/spec,verbs=get
+func (r *ReconcileCloudCredentialConfig) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithFields(log.Fields{
+		"controller": controllerName,
+		"cr":         fmt.Sprintf("%s/%s", request.NamespacedName.Namespace, request.NamespacedName.Name),
+	})
+
+	// Check OperatorLogLevel and LogLevel
+	log.Debug("Checking log level for Reconcile request")
+	desiredOperatorLogLevel, err := utils.GetOperatorLogLevel(r, logger)
+	if err != nil || !logLevelUtils.ValidLogLevel(desiredOperatorLogLevel) {
+		return reconcile.Result{Requeue: false}, fmt.Errorf("the operatorLogLevel defined for the cloudcredential object is invalid")
+	}
+
+	desiredLogLevel, err := utils.GetLogLevel(r, logger)
+	if err != nil || !logLevelUtils.ValidLogLevel(desiredLogLevel) {
+		return reconcile.Result{Requeue: false}, fmt.Errorf("the logLevel defined for the cloudcredential object is invalid")
+	}
+
+	// Pick the logLevel with the highest verbosity level.
+	maxLogLevel := desiredOperatorLogLevel
+	if logLevelUtils.LogLevelToVerbosity(desiredLogLevel) > logLevelUtils.LogLevelToVerbosity(desiredOperatorLogLevel) {
+		maxLogLevel = desiredLogLevel
+	}
+
+	currentLogLevel, errBool := logLevelUtils.GetLogLevel()
+	if errBool {
+		log.Debug("Unable to get the current loglevel. Defaulting the current level to 'Normal'")
+	}
+
+	if currentLogLevel == maxLogLevel {
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	// Set logLevel to the new level
+	err = logLevelUtils.SetLogLEvel(maxLogLevel)
+	if err != nil {
+		log.Errorf("Unable to set loglevel. %v", err)
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	log.Infof("Log level changed to %v", maxLogLevel)
+
```

Comment:
This a little tricky for our operator. We use logrus throughout, and the logLevelUtils.GetLogLevel() returns the level for klog.
I would propose dropping this detection logic and just blindly assigning/re-assigning the log level based on a change the CloudCredentials CR.

And because we use logrus, we need to handle setting the log level for that package as well.

Something like this:

```suggestion
	// Set klog logLevel to the desired level
	err = logLevelUtils.SetLogLEvel(maxLogLevel)
	if err != nil {
		log.Errorf("Unable to set loglevel. %v", err)
		return reconcile.Result{Requeue: false}, nil
	}

	log.Infof("klog level changed to %v", maxLogLevel)
	
	var logrusLevel log.Level
	switch maxLogLevel {
	case operatorv1.Debug:
	        logrusLevel = log.DebugLevel
	case operatorv1.Trace, operatorv1.TraceAll:
	        logrusLevel = log.TraceLevel
	case operatorv1.Normal:
	       logrusLevel = log.InfoLevel
	 default:
	         logrusLevel = log.InfoLevel
	  }
	  
	  log.SetLevel(logrusLevel)
	  log.Infof("logrus level changed to %s", logrusLevel)

```

Reply:
I have made these changes to include the logrus level changes.
https://github.com/openshift/cloud-credential-operator/pull/379/files#diff-e983e87892f7cdbd2b02917acb48aa087f3a4955ddf91b27354a2ce0dcb0dcbfR125-R153

I did keep the early exit clause:
```
	if currentLogLevel == maxLogLevel {
		return reconcile.Result{Requeue: false}, nil
	}
```

If you think this is wrong, let me know and I can fix it up :)